app/controllers/schemes_controller.rb

```diff
@@ -251,12 +272,55 @@ def find_resource
   def authenticate_scope!
     head :unauthorized and return unless current_user.data_coordinator? || current_user.support?
 
-    if %w[show locations primary_client_group confirm_secondary_client_group secondary_client_group support details check_answers edit_name].include?(action_name) && !((current_user.organisation == @scheme&.owning_organisation) || current_user.support?)
+    if %w[show locations primary_client_group confirm_secondary_client_group secondary_client_group support details check_answers edit_name deactivate].include?(action_name) && !((current_user.organisation == @scheme&.owning_organisation) || current_user.support?)
       render_not_found and return
     end
   end
 
   def redirect_if_scheme_confirmed
     redirect_to @scheme if @scheme.confirmed?
   end
+
+  def confirm_deactivation
+    if @scheme.update(deactivation_date: params[:scheme][:deactivation_date])
+      flash[:notice] = "#{@scheme.service_name} has been deactivated"
+    end
+    redirect_to scheme_details_path(@scheme)
+  end
+
+  def deactivation_date_errors
+    if params[:scheme][:deactivation_date].blank? && params[:scheme][:deactivation_date_type].blank?
+      @scheme.errors.add(:deactivation_date_type, message: I18n.t("validations.scheme.deactivation_date.not_selected"))
+    end
+
+    if params[:scheme][:deactivation_date_type] == "other"
+      day = params[:scheme]["deactivation_date(3i)"]
+      month = params[:scheme]["deactivation_date(2i)"]
+      year = params[:scheme]["deactivation_date(1i)"]
+
+      collection_start_date = FormHandler.instance.current_collection_start_date
+
+      if [day, month, year].any?(&:blank?)
+        @scheme.errors.add(:deactivation_date, message: I18n.t("validations.scheme.deactivation_date.not_entered"))
+      elsif !Date.valid_date?(year.to_i, month.to_i, day.to_i)
+        @scheme.errors.add(:deactivation_date, message: I18n.t("validations.scheme.deactivation_date.invalid"))
+      elsif !Date.new(year.to_i, month.to_i, day.to_i).between?(collection_start_date, Date.new(2200, 1, 1))
+        @scheme.errors.add(:deactivation_date, message: I18n.t("validations.scheme.deactivation_date.out_of_range", date: collection_start_date.to_formatted_s(:govuk_date)))
+      end
+    end
```

Comment:
These validations are likely better suited to the model. Was there a reason for having them here?

Reply:
Nope, I'll move them across 👍